### PR TITLE
[fsdp,algo] feat: add use_length_grouped_bsz to reduce padding waste in FSDP training

### PR DESCRIPTION
**Pull request by roeeh21**
## Summary

For models that cannot use sequence packing (e.g. Mamba/SSM), training under FSDP wastes compute on padded tokens. This PR adds an opt-in `use_length_grouped_bsz` option that mitigates that waste by greedily grouping sequences of similar length into micro-batches.

(See [blog post](https://www.ai21.com/blog/padding-minimization-efficiency/) for the broader motivation behind padding-aware micro-batching.)

## Approach

When `actor_rollout_ref.model.use_length_grouped_bsz=True`:

- Sort sequences in the batch by length (descending).
- Greedily pack them into micro-batches such that `len(micro_batch) * max_seq_in_micro_batch <= max_token_len_per_gpu`.
- All-reduce the micro-batch count across DP ranks (split the largest batches on lagging ranks) so the optimizer step stays synchronized.

Off by default. Mutually exclusive with `use_remove_padding` and Ulysses SP.

## Files

- `verl/utils/seqlen_balancing.py`: new helper `get_length_grouped_micro_batches()` returning `(micro_batches, indices)`, mirroring the contract of `rearrange_micro_batches`.
- `verl/workers/engine/utils.py`: dispatcher branch for `use_length_grouped_bsz` in `prepare_micro_batches`.
- `verl/workers/engine/fsdp/transformer_impl.py`: validation guards (mutually exclusive with `use_remove_padding` and Ulysses SP).
- `verl/workers/engine_workers.py`: forwards the flag through to training and inference data.
- Configs: new `use_length_grouped_bsz` field in model + actor configs (YAML + dataclass), plus the regenerated `_generated_ppo_*` yamls.

## Benchmarks

10-step GRPO on Qwen 2.5 3B-Instruct / GSM8K / 1x4 H100. Mean of `timing_s/update_actor` over steps 2-10:

| Config | mean update_actor (s) |
|---|---|
| `use_remove_padding=True` (rmpad) | 2.235 |
| **`use_length_grouped_bsz=True`** | **2.475** |
| dense (no rmpad, no dynamic_bsz) | 11.692 |

`use_length_grouped_bsz` is **~4.7x faster than dense** and **~10.7% slower than `use_remove_padding`**.

## Test Plan

- [x] 10-step GRPO run with `use_length_grouped_bsz=True` completes end-to-end on Qwen 2.5 3B / GSM8K / 1x4 H100.
- [x] Comparison runs with `use_remove_padding` and dense-padding baselines reproduce the speedup.

This PR was developed with AI assistance.

### CODE REVIEW

File: verl/workers/actor/dp_actor.py [hunk 1]
```diff
@@ -343,43 +358,84 @@ def _forward_micro_batch(
...
+                response_start_idx = torch.full((batch_size,), seqlen - response_length - 1, device=input_ids.device)
+                response_end_idx = torch.full((batch_size,), seqlen - 1, device=input_ids.device)
+
+                if self.truncate_padding:
+                    # Roll left padding to the right so real tokens are contiguous on the left
+                    left_padding = attention_mask.argmax(dim=1)
+                    for key in model_inputs:
+                        rolled = torch.empty_like(model_inputs[key])
+                        for i in range(batch_size):
+                            rolled[i] = torch.roll(model_inputs[key][i], -int(left_padding[i]), dims=-1)
```
Comment (asafkar) [inline verl/workers/actor/dp_actor.py]: `torch.roll` creates a copy of a tensor, and is usually slower than the alternative, 
which is manual slicing and `torch.cat`. 
Generally I think at least one of these for-loops can be removed and a vectorized code be implemented instead.

File: verl/workers/actor/dp_actor.py [hunk 2]
```diff
@@ -343,43 +358,84 @@ def _forward_micro_batch(
...
+                        model_inputs[key] = rolled
+                    response_start_idx -= left_padding
+                    response_end_idx -= left_padding
+
+                    # Trim common right padding (all padding is on the right after rolling)
+                    common_pad_right = model_inputs["attention_mask"].flip(dims=[-1]).argmax(dim=-1).min().item()
```
Comment (asafkar) [inline verl/workers/actor/dp_actor.py]: `flip` can also be optimized, by using `torch.where` with a mask

File: verl/utils/seqlen_balancing.py [hunk 3]
```diff
@@ -345,6 +348,159 @@ def roundup_divisible(a: int, b: int) -> int:
...
+        return micro_batches_idx
+
+    num_micro_batches = len(micro_batches_idx)
+    num_micro_batches_tensor = torch.tensor([num_micro_batches], device=get_device_name())
+    dist.all_reduce(num_micro_batches_tensor, op=dist.ReduceOp.MAX, group=dp_group)
+    max_num_micro_batches = num_micro_batches_tensor.cpu().item()
+
+    while len(micro_batches_idx) < max_num_micro_batches:
+        largest_batch_idx = max(range(len(micro_batches_idx)), key=lambda i: len(micro_batches_idx[i]))
+        largest_batch = micro_batches_idx[largest_batch_idx]
```
[the remark this answers is not in the document]
Comment (roeeh21) [reply verl/utils/seqlen_balancing.py]: Fixed in commit def9fcfb — added an explicit guard that raises a clear RuntimeError when a rank has an empty batch but other ranks don't.

File: verl/utils/seqlen_balancing.py [hunk 4]
```diff
@@ -345,6 +348,159 @@ def roundup_divisible(a: int, b: int) -> int:
...
+
+    sorted_sequence_lengths_with_idx = sorted(
+        [(length, idx) for idx, length in enumerate(sequence_lengths)], key=lambda x: x[0], reverse=True
+    )
+
+    micro_batches_idx = []
+
+    longest_sequence_length, longest_sequence_idx = sorted_sequence_lengths_with_idx[0]
+    current_micro_batch_idx = [longest_sequence_idx]
+    current_micro_batch_max_len = longest_sequence_length
```
[the remark this answers is not in the document]
Comment (roeeh21) [reply verl/utils/seqlen_balancing.py]: Fixed in commit def9fcfb — added an early return for empty input batches.

File: verl/workers/actor/dp_actor.py [hunk 5]
```diff
@@ -343,43 +358,84 @@ def _forward_micro_batch(
...
+                response_end_idx = torch.full((batch_size,), seqlen - 1, device=input_ids.device)
+
+                if self.truncate_padding:
+                    # Roll left padding to the right so real tokens are contiguous on the left
+                    left_padding = attention_mask.argmax(dim=1)
+                    for key in model_inputs:
+                        rolled = torch.empty_like(model_inputs[key])
+                        for i in range(batch_size):
+                            rolled[i] = torch.roll(model_inputs[key][i], -int(left_padding[i]), dims=0)
+                        model_inputs[key] = rolled
```
[the remark this answers is not in the document]
Comment (roeeh21) [reply verl/workers/actor/dp_actor.py]: Fixed in commit b621e234 — replaced the per-sequence Python loop + `torch.roll` with a batched `torch.gather` using precomputed indices:

```python
shifts = left_padding.long()
arange_l = torch.arange(seqlen, device=input_ids.device)
roll_indices = (arange_l + shifts.unsqueeze(1)) % seqlen
```

This handles 2D `(B, L)` and 3D `(C, B, L)` / `(B, C, L)` tensors correctly by gathering along the appropriate dimension, with no Python loop.

File: verl/workers/actor/dp_actor.py (hunk 1 above)
Comment (roeeh21) [reply verl/workers/actor/dp_actor.py]: Good call — replaced with a vectorized `torch.gather` approach in b621e234. No Python loop, single kernel launch:

```python
shifts = left_padding.long()
arange_l = torch.arange(seqlen, device=input_ids.device)
roll_indices = (arange_l + shifts.unsqueeze(1)) % seqlen
# then torch.gather per dimension
```

This also properly handles 3D position_ids (e.g. qwen2vl mrope) by expanding indices along the correct dimension.

File: verl/workers/actor/dp_actor.py (hunk 2 above)
Comment (roeeh21) [reply verl/workers/actor/dp_actor.py]: Done in b621e234 — replaced `flip(dims=[-1]).argmax()` with:

```python
last_one_idx = (mask * torch.arange(seqlen, device=mask.device)).max(dim=-1)[0]
common_pad_right = (seqlen - 1 - last_one_idx).min().item()
```

Avoids the flip copy entirely.

File: verl/workers/actor/dp_actor.py [hunk 6]
```diff
@@ -343,43 +360,110 @@ def _forward_micro_batch(
...
+                model_inputs = {
+                    "input_ids": input_ids,
+                    "attention_mask": attention_mask,
+                    "position_ids": position_ids,
+                }
+
+                response_start_idx = torch.full((batch_size,), seqlen - response_length - 1, device=input_ids.device)
+                response_end_idx = torch.full((batch_size,), seqlen - 1, device=input_ids.device)
+
+                if self.truncate_padding:
```
Comment (ai21shakedmeirom) [inline verl/workers/actor/dp_actor.py]: Please add a general comment here, explaining what process is going to take place - truncating the padding to the possible minimum (in addition to each separate comment you added about the left to right and about the trimming)

File: verl/workers/actor/dp_actor.py [hunk 7]
```diff
@@ -343,43 +360,110 @@ def _forward_micro_batch(
...
+                    # Roll left padding to the right so real tokens are contiguous on the left
+                    left_padding = attention_mask.argmax(dim=1)
+                    shifts = left_padding.long()
+                    arange_l = torch.arange(seqlen, device=input_ids.device)
+                    roll_indices = (arange_l + shifts.unsqueeze(1)) % seqlen  # (B, L)
+                    for key, t in list(model_inputs.items()):
+                        if t.dim() == 2:
```
Comment (ai21shakedmeirom) [inline verl/workers/actor/dp_actor.py]: Please add a comment for each of these "if"s stating what is the expected situation (e.g. for `t.dim()==2` what these 2 dims are supposed to be, and then in the later "elif" add comment about what are the 3 dims suppose to be...)

**Comment by wuxibin89** [comment]
@roeeh21 Hi, thanks for you contribution. We will deprecate legacy engine([fsdp_workers.py](https://github.com/verl-project/verl/blob/main/verl/workers/fsdp_workers.py)/[megatron_workers.py](https://github.com/verl-project/verl/blob/main/verl/workers/megatron_workers.py)) and switch to new engine [engine_workers.py](https://github.com/verl-project/verl/blob/main/verl/workers/engine_workers.py) soon.

Could you also help to support this feature in new [fsdp engine](https://github.com/verl-project/verl/blob/main/verl/workers/engine/fsdp/transformer_impl.py)? You can enable it by `trainer.use_legacy_worker_impl=disable`.

**Comment by roeeh21** [comment]
@wuxibin89 Done! The latest push adds full support for the new engine (`engine_workers.py` / `FSDPEngine`), activated with `trainer.use_legacy_worker_impl=disable`.

Changes for the new engine path:
- `verl/workers/engine/utils.py`: truncate_padding branch in `prepare_micro_batches` using jagged tensor greedy packing
- `verl/workers/engine/fsdp/transformer_impl.py`: config wiring + validation guards (mutually exclusive with `use_remove_padding` and Ulysses SP)
- `verl/workers/engine_workers.py`: passes `truncate_padding` flag into training/inference data
- `verl/utils/seqlen_balancing.py`: `get_truncate_padding_micro_batches_jagged` for nested tensor packing with cross-rank sync

Validated with 50-step runs on Qwen2.5-3B + GSM8K (8xH100). Truncate padding is ~2x faster than dense and within ~15% of remove_padding on actor update time. See updated PR description for full benchmark.

**Comment by roeeh21** [comment]
@wuxibin89 Rebased onto latest main. New engine support is included. Would appreciate a review!

**Comment by roeeh21** [comment]
Gentle ping @wuxibin89 @eric-haibin-lin @vermouth1992 @tongyx361 @PeterSH6, merged latest main to bring the branch up to date. Would appreciate a review when you have a moment, happy to address any feedback.

**Comment by wuxibin89** [comment]
@roeeh21 Thanks for your contribution, the legacy worker has been removed, please rebase main to resolve the conflicts.

**Comment by roeeh21** [comment]
@wuxibin89 PR has been updated as requested, ready for review. Thanks!
